core/server/src/main/java/org/phenotips/remote/server/internal/IncomingSearchRequestProcessor.java

```diff
@@ -96,16 +97,19 @@ public JSONObject processHTTPSearchRequest(ApiDataConverter apiVersionSpecificCo
             return responseJSON;
         } catch (JSONException ex) {
             this.logger.error("Incorrect incoming request: misformatted JSON");
+            // save raw request data for audit purposes only
+            this.saveUnprocessedRequest(stringJson, remoteServerId, apiVersionSpecificConverter.getApiVersion());
             return apiVersionSpecificConverter.generateWrongInputDataResponse("misformatted JSON");
         } catch (ApiViolationException ex) {
             this.logger.error("Error converting JSON to incoming request");
+            // save raw request data for audit purposes only
+            this.saveUnprocessedRequest(stringJson, remoteServerId, apiVersionSpecificConverter.getApiVersion());
             return apiVersionSpecificConverter.generateWrongInputDataResponse(ex.getMessage());
         } catch (Exception ex) {
             this.logger.error("CODE Error: {}", ex);
-            return apiVersionSpecificConverter.generateInternalServerErrorResponse(null);
-        } finally {
             // save raw request data for audit purposes only
             this.saveUnprocessedRequest(stringJson, remoteServerId, apiVersionSpecificConverter.getApiVersion());
+            return apiVersionSpecificConverter.generateInternalServerErrorResponse(null);
```

Comment:
Hmm, there must be a solution with less code duplication. For example, a "requestIncomplete" variable, initially initialised to "true", and set to false right before the last return statement in the normal code flow. Then in `finally` we do `if (requestIncomplete) { .... }`